craft_application/services/proxy.py

```diff
@@ -0,0 +1,165 @@
+
+        :param proxy_cert: The path to the proxy certificate to install in the instance.
+        :param http_proxy: The proxy url to set in the instance.
+        """
+        self.__proxy_cert = proxy_cert
+        self.__http_proxy = http_proxy
+        self.__is_configured = True
+
+    @final
+    def configure_instance(self, instance: craft_providers.Executor) -> dict[str, str]:
+        """Configure a build instance to connect to a proxy.
+
+        :param instance: The instance to configure.
+
+        :returns: A dict of environment variables to set in the instance.
+        """
+        if not self.__is_configured:
+            emit.debug(
+                "Skipping proxy configuration because the proxy service isn't configured."
+            )
+            return {}
+
+        emit.progress("Configuring proxy in instance.")
+
+        self._install_certificate(instance)
+        self._configure_pip(instance)
+        self._configure_snapd(instance)
+        self._configure_apt(instance)
+
+        return self._env
+
+    @property
+    def _env(self) -> dict[str, str]:
+        """Environment variables to use for the proxy."""
+        return {
+            "http_proxy": self.__http_proxy,
+            "https_proxy": self.__http_proxy,
+            # This makes the requests lib take our cert into account.
+            "REQUESTS_CA_BUNDLE": str(_PROXY_CERT_INSTANCE_PATH),
+            # Same, but for cargo.
+            "CARGO_HTTP_CAINFO": str(_PROXY_CERT_INSTANCE_PATH),
+            # Have go download directly from repositories
+            "GOPROXY": "direct",
+        }
+
+    def _configure_pip(self, instance: craft_providers.Executor) -> None:
+        emit.progress("Configuring pip")
+
+        self._execute_run(instance, ["mkdir", "-p", "/root/.pip"])
+        pip_config = b"[global]\ncert=/usr/local/share/ca-certificates/local-ca.crt"
+        instance.push_file_io(
+            destination=pathlib.Path("/root/.pip/pip.conf"),
+            content=io.BytesIO(pip_config),
+            file_mode="0644",
+        )
+
+    def _configure_snapd(self, instance: craft_providers.Executor) -> None:
+        """Configure snapd to use the proxy and see our certificate.
+
+        Note: This must be called after _install_certificate(), to ensure that
+        when the snapd restart happens the new cert is there.
+        """
+        emit.progress("Configuring snapd")
+        self._execute_run(instance, ["systemctl", "restart", "snapd"])
+        for config in ("proxy.http", "proxy.https"):
+            self._execute_run(
+                instance, ["snap", "set", "system", f"{config}={self.__http_proxy}"]
+            )
+
+    def _configure_apt(self, instance: craft_providers.Executor) -> None:
+        emit.progress("Configuring Apt")
+        apt_config = f'Acquire::http::Proxy "{self.__http_proxy}";\n'
+        apt_config += f'Acquire::https::Proxy "{self.__http_proxy}";\n'
+
+        instance.push_file_io(
```

Comment:
How does this react with an Alma Linux instance? Should we check for `/etc/apt` and skip configuring if that doesn't exist?

(Making yum/dnf/pacman/emerge/apk respect this proxy is out of scope)

Reply:
I'm not sure but enabling the fetch service for an Alma instance should fail the same way with or without this PR. Can we add safety rails in a follow-up PR?

Reply:
The issues is that we're now always configuring the proxy service when we enable a provider.

Reply:
[It was pointed out to me elsewhere that we're no-opping right now if the fetch service doesn't configure the proxy service]